### PR TITLE
patool: fix the non-use of hard-coded binary

### PR DESCRIPTION
**Pull request by marius851000**
###### Motivation for this change

https://gitlab.com/portmod/portmod/-/issues/327

(didn't used the hard-coded function at all)

###### Things done

Simple fix. The function didn't did what I expected. Just use $PATH + ":/nix/store/.../bin:/nix/store/.../bin:...". Tested with both existing and empty path.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/patool/default.nix [hunk 1]
```diff
@@ -36,7 +36,7 @@ buildPythonPackage rec {
 
   prePatch = ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/patool/default.nix:37]:
```suggestion
  postPatch = ''
```

Comment (marius851000) [reply pkgs/development/python-modules/patool/default.nix]: Why is prePatch better than postPatch ? Changed it anyway.

**Comment by symphorien** [comment]
I can extract .tar.bz2 and .tar.xz with the patool version of current nixos-unstable. What does this fix?

**Comment by marius851000** [comment]
Patool should have used the binary in your PATH. You can check this by running `patool formats`.

**Comment by symphorien** [comment]
With this PR, patool still complains that gzip is missing:
```
$  PATH= /nix/store/4hpn3lyasli5hdin3nk5a28vn6qaf2n3-python3.9-patool-1.12/bin/patool formats
Archive programs of patool 1.12
Archive programs are searched in the following directories:
/nix/store/fkzla307l4mlcvfyshsrccwl7szihx2z-python3-3.9.6/bin:/nix/store/4hpn3lyasli5hdin3nk5a28vn6qaf2n3-python3.9-patool-1.12/bin

7z files:
       list: - (no program found; install 7z, 7za or 7zr)
    extract: - (no program found; install 7z, 7za or 7zr)
       test: - (no program found; install 7z, 7za or 7zr)
     create: - (no program found; install 7z, 7za or 7zr)
ace files:
       list: - (no program found; install unace)
    extract: - (no program found; install unace)
       test: - (no program found; install unace)
     create: - (not supported)
adf files:
       list: - (no program found; install unadf)
    extract: - (no program found; install unadf)
       test: - (no program found; install unadf)
     create: - (not supported)
alzip files:
       list: - (no program found; install unalz)
    extract: - (no program found; install unalz)
       test: - (no program found; install unalz)
     create: - (not supported)
ape files:
       list: py_echo 
    extract: - (no program found; install mac)
       test: - (no program found; install mac)
     create: - (no program found; install mac)
ar files:
       list: - (no program found; install ar)
    extract: - (no program found; install ar)
       test: - (no program found; install ar)
     create: - (no program found; install ar)
arc files:
       list: - (no program found; install arc)
    extract: - (no program found; install arc)
       test: - (no program found; install arc)
     create: - (no program found; install arc)
arj files:
       list: - (no program found; install arj)
    extract: - (no program found; install arj)
       test: - (no program found; install arj)
     create: - (no program found; install arj)
bzip2 files:
       list: - (no program found; install 7z or 7za)
    extract: py_bz2 
       test: - (no program found; install 7z or 7za)
     create: py_bz2 
cab files:
       list: - (no program found; install cabextract or 7z)
    extract: - (no program found; install cabextract or 7z)
       test: - (no program found; install cabextract or 7z)
     create: - (no program found; install lcab)
chm files:
       list: - (not supported)
    extract: - (no program found; install archmage or extract_chmLib)
       test: - (no program found; install archmage)
     create: - (not supported)
compress files:
       list: py_echo 
    extract: - (no program found; install gzip, 7z, 7za or uncompress.real)
       test: - (no program found; install gzip, 7z or 7za)
     create: - (no program found; install compress)
cpio files:
       list: - (no program found; install cpio, bsdcpio or 7z)
    extract: - (no program found; install cpio, bsdcpio or 7z)
       test: - (no program found; install cpio, bsdcpio or 7z)
     create: - (no program found; install cpio or bsdcpio)
deb files:
       list: - (no program found; install dpkg-deb or 7z)
    extract: - (no program found; install dpkg-deb or 7z)
       test: - (no program found; install dpkg-deb or 7z)
     create: - (not supported)
dms files:
       list: - (no program found; install xdms)
    extract: - (no program found; install xdms)
       test: - (no program found; install xdms)
     create: - (not supported)
flac files:
       list: py_echo 
    extract: - (no program found; install flac)
       test: - (no program found; install flac)
     create: - (no program found; install flac)
gzip files:
       list: - (no program found; install 7z, 7za, pigz or gzip)
    extract: py_gzip 
       test: - (no program found; install 7z, 7za, pigz or gzip)
     create: py_gzip 
iso files:
       list: - (no program found; install 7z or isoinfo)
    extract: - (no program found; install 7z)
       test: - (no program found; install 7z)
     create: - (no program found; install genisoimage)
lrzip files:
       list: py_echo 
    extract: - (no program found; install lrzip)
       test: - (no program found; install lrzip)
     create: - (no program found; install lrzip)
lzh files:
       list: - (no program found; install lha)
    extract: - (no program found; install lha)
       test: - (no program found; install lha)
     create: - (no program found; install lha)
lzip files:
       list: py_echo 
    extract: - (no program found; install plzip, lzip, clzip or pdlzip)
       test: - (no program found; install plzip, lzip, clzip or pdlzip)
     create: - (no program found; install plzip, lzip, clzip or pdlzip)
lzma files:
       list: py_echo 
    extract: py_lzma 
       test: - (no program found; install 7z, lzma or xz)
     create: py_lzma 
lzop files:
       list: - (no program found; install lzop)
    extract: - (no program found; install lzop)
       test: - (no program found; install lzop)
     create: - (no program found; install lzop)
rar files:
       list: - (no program found; install rar)
    extract: - (no program found; install rar)
       test: - (no program found; install rar)
     create: - (no program found; install rar)
rpm files:
       list: - (no program found; install rpm, 7z or 7za)
    extract: - (no program found; install rpm2cpio or 7z)
       test: - (no program found; install rpm or 7z)
     create: - (not supported)
rzip files:
       list: py_echo 
    extract: - (no program found; install rzip)
       test: - (not supported)
     create: - (no program found; install rzip)
shar files:
       list: - (not supported)
    extract: - (no program found; install unshar)
       test: - (not supported)
     create: - (no program found; install shar)
shn files:
       list: py_echo 
    extract: - (no program found; install shorten)
       test: - (not supported)
     create: - (no program found; install shorten)
tar files:
       list: py_tarfile 
    extract: py_tarfile 
       test: py_tarfile 
     create: py_tarfile 
vhd files:
       list: - (no program found; install 7z)
    extract: - (no program found; install 7z)
       test: - (no program found; install 7z)
     create: - (not supported)
xz files:
       list: - (no program found; install xz or 7z)
    extract: py_lzma 
       test: - (no program found; install xz or 7z)
     create: py_lzma 
zip files:
       list: py_zipfile 
    extract: py_zipfile 
       test: py_zipfile 
     create: py_zipfile 
zoo files:
       list: - (no program found; install zoo)
    extract: - (no program found; install zoo)
       test: - (no program found; install zoo)
     create: - (no program found; install zoo)
zpaq files:
       list: - (no program found; install zpaq)
    extract: - (no program found; install zpaq)
       test: - (no program found; install zpaq)
     create: - (no program found; install zpaq)
```

**Comment by marius851000** [comment]
That's not normal. When running 
1. `git checkout 9012f0c8f1ff7f5c445fe3f198d4ef6992dec5fc`
2. `export PATH=result/bin`
3. `patool formats`

it is still able to locate gzip on my side :
```
gzip files:
       list: /nix/store/wdalgxw1447kfpkgcz8gisq5vm8splhy-p7zip-17.04/bin/7z 
    extract: /nix/store/wdalgxw1447kfpkgcz8gisq5vm8splhy-p7zip-17.04/bin/7z 
       test: /nix/store/wdalgxw1447kfpkgcz8gisq5vm8splhy-p7zip-17.04/bin/7z 
     create: /nix/store/wdalgxw1447kfpkgcz8gisq5vm8splhy-p7zip-17.04/bin/7z 
```

notice that `ls -lAh result` return `/nix/store/97wnngraqxbksbwgc2pl5fw1w1cav3xr-python3.9-patool-1.12` and not your `/nix/store/4hpn3lyasli5hdin3nk5a28vn6qaf2n3-python3.9-patool-1.12/bin/patool` . I'll try to rebase on master, maybe it'll help to see the error.

**Comment by symphorien** [comment]
With your commit I get the same hash and the correct result. The hash I got was in nixpkgs-review, which merges on master, but now if I retry I get the same hash as I got before and the correct result. I'm very puzzled...

**Comment by marius851000** [comment]
I did the test myself, and it indeed seems to be due to an update in nixpkgs. I'll take a look at this.

**Comment by marius851000** [comment]
Rebased again on master. The issue doesn't seems to happen anymore. Should be good to merge.

File: pkgs/development/python-modules/patool/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/patool/default.nix]: We want to change the file after patches has been applied. Every package that uses prePatch like this potentially breaks patches.

**Comment by Artturin** [comment]
please verify that the following is correct

```bash
[nix-shell:~/.cache/nixpkgs-review/pr-145130]$ PATH="" /nix/store/rx6a33kkr6qj05rjx0c3ss0xa327j07x-python3.9-pat[89/310]
/bin/patool formats
Archive programs of patool 1.12
Archive programs are searched in the following directories:
/nix/store/33hlmb4afkk7l2fapfik80m6yknjxq10-python3-3.9.12/bin:/nix/store/rx6a33kkr6qj05rjx0c3ss0xa327j07x-python3.9-pat
ool-1.12/bin

7z files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z (rar archives not supported)
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z (rar archives not supported)
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z (rar archives not supported)
     create: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z (rar archives not supported)
ace files:
       list: - (no program found; install unace)
    extract: - (no program found; install unace)
       test: - (no program found; install unace)
     create: - (not supported)
adf files:
       list: - (no program found; install unadf)
    extract: - (no program found; install unadf)
       test: - (no program found; install unadf)
     create: - (not supported)
alzip files:
       list: - (no program found; install unalz)
    extract: - (no program found; install unalz)
       test: - (no program found; install unalz)
     create: - (not supported)
ape files:
       list: py_echo
    extract: - (no program found; install mac)
       test: - (no program found; install mac)
     create: - (no program found; install mac)
ar files:
       list: - (no program found; install ar)
    extract: - (no program found; install ar)
       test: - (no program found; install ar)
     create: - (no program found; install ar)
arc files:
       list: - (no program found; install arc)
    extract: - (no program found; install arc)
       test: - (no program found; install arc)
     create: - (no program found; install arc)
arj files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z     create: - (no program found; install arj)
bzip2 files:       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7zcab files:
       list: /nix/store/gjy43l9sp02819yzfcf5g0maarah8jvc-cabextract-1.9.1/bin/cabextract
    extract: /nix/store/gjy43l9sp02819yzfcf5g0maarah8jvc-cabextract-1.9.1/bin/cabextract
       test: /nix/store/gjy43l9sp02819yzfcf5g0maarah8jvc-cabextract-1.9.1/bin/cabextract
     create: - (no program found; install lcab)
chm files:
       list: - (not supported)
    extract: - (no program found; install archmage or extract_chmLib)
       test: - (no program found; install archmage)
     create: - (not supported)
compress files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/13y9c8f2jxwmszbnlp86qrzb1bvyl95a-gzip-1.12/bin/gzip
       test: /nix/store/13y9c8f2jxwmszbnlp86qrzb1bvyl95a-gzip-1.12/bin/gzip
     create: - (no program found; install compress)
cpio files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: - (no program found; install cpio or bsdcpio)
deb files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: - (not supported)
dms files:
       list: - (no program found; install xdms)
    extract: - (no program found; install xdms)
       test: - (no program found; install xdms)
     create: - (not supported)
flac files:
       list: py_echo
    extract: - (no program found; install flac)
       test: - (no program found; install flac)
     create: - (no program found; install flac)
gzip files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
iso files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: - (no program found; install genisoimage)
lrzip files:
       list: py_echo
    extract: - (no program found; install lrzip)
       test: - (no program found; install lrzip)
     create: - (no program found; install lrzip)
lzh files:
       list: - (no program found; install lha)
    extract: - (no program found; install lha)
       test: - (no program found; install lha)
     create: - (no program found; install lha)
lzip files:
       list: py_echo
    extract: /nix/store/ygmnlmf8hwsdm2pgf27rpm1b5lvhzra7-lzip-1.22/bin/lzip
       test: /nix/store/ygmnlmf8hwsdm2pgf27rpm1b5lvhzra7-lzip-1.22/bin/lzip
     create: /nix/store/ygmnlmf8hwsdm2pgf27rpm1b5lvhzra7-lzip-1.22/bin/lzip
lzma files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: /nix/store/0fnnx443l6107v99719vd9n04wrznmas-xz-5.2.5-bin/bin/lzma
lzop files:
       list: - (no program found; install lzop)
    extract: - (no program found; install lzop)
       test: - (no program found; install lzop)
     create: - (no program found; install lzop)
rar files:
       list: - (no program found; install rar)
    extract: - (no program found; install rar)
       test: - (no program found; install rar)
     create: - (no program found; install rar)
rpm files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: - (not supported)
rzip files:
       list: py_echo
    extract: - (no program found; install rzip)
       test: - (not supported)
     create: - (no program found; install rzip)
shar files:
       list: - (not supported)
    extract: - (no program found; install unshar)
       test: - (not supported)
     create: - (no program found; install shar)
shn files:
       list: py_echo
    extract: - (no program found; install shorten)
       test: - (not supported)
     create: - (no program found; install shorten)
tar files:
       list: /nix/store/7dc5r0a6xy7mglbi57hmhvcwm1nqjwiz-gnutar-1.34/bin/tar (supported compressions: bzip2, gzip, lzip,
 lzma, xz)
    extract: /nix/store/7dc5r0a6xy7mglbi57hmhvcwm1nqjwiz-gnutar-1.34/bin/tar (supported compressions: bzip2, gzip, lzip,
 lzma, xz)
       test: /nix/store/7dc5r0a6xy7mglbi57hmhvcwm1nqjwiz-gnutar-1.34/bin/tar (supported compressions: bzip2, gzip, lzip,
 lzma, xz)
     create: /nix/store/7dc5r0a6xy7mglbi57hmhvcwm1nqjwiz-gnutar-1.34/bin/tar (supported compressions: bzip2, gzip, lzip,
 lzma, xz)
vhd files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: - (not supported)
xz files:
       list: /nix/store/0fnnx443l6107v99719vd9n04wrznmas-xz-5.2.5-bin/bin/xz
    extract: /nix/store/0fnnx443l6107v99719vd9n04wrznmas-xz-5.2.5-bin/bin/xz
       test: /nix/store/0fnnx443l6107v99719vd9n04wrznmas-xz-5.2.5-bin/bin/xz
     create: /nix/store/0fnnx443l6107v99719vd9n04wrznmas-xz-5.2.5-bin/bin/xz
zip files:
       list: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
    extract: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
       test: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
     create: /nix/store/zbblc1dskij305ff4bn1n5fbminx87h6-p7zip-17.04/bin/7z
zoo files:
       list: - (no program found; install zoo)
    extract: - (no program found; install zoo)
       test: - (no program found; install zoo)
     create: - (no program found; install zoo)
zpaq files:
       list: /nix/store/9g82129l56ygx1qwgi13hr7yxdp2ri0i-zpaq-7.15/bin/zpaq
    extract: /nix/store/9g82129l56ygx1qwgi13hr7yxdp2ri0i-zpaq-7.15/bin/zpaq
       test: /nix/store/9g82129l56ygx1qwgi13hr7yxdp2ri0i-zpaq-7.15/bin/zpaq
     create: /nix/store/9g82129l56ygx1qwgi13hr7yxdp2ri0i-zpaq-7.15/bin/zpaq

**Comment by marius851000** [comment]
Removed zopfli and unzip

**Comment by K900** [comment]
@ofborg build patool

**Pull request merged**